### PR TITLE
Add shell completion support for CLI

### LINKED ISSUE #887: Autocompletion for the CLI using tab
It would be great if we can using `tab` to auto-complete the cluster name when typing the CLI commands. I tried to implement that in [cluster-auto-complete](https://github.com/sky-proj/sky/tree/cluster-auto-complete), but the auto-completion is very laggy at that time.

### PR DESCRIPTION
**Pull request by iojw**
This adds support for [shell completion using click](https://click.palletsprojects.com/en/8.1.x/shell-completion/). Under the hood, the CLI runs SQL queries against the local data store to figure out what cluster names can be used to autocomplete the command.

To enable shell completion, the user needs to modify their shell configuration as discussed [here](https://click.palletsprojects.com/en/8.1.x/shell-completion/). To simplify this process, I added a `--install-shell-completion` option to the CLI that automates this process for a specific shell. For example, to enable shell completion for `zsh`, the user will need to run `sky --install-shell-completion zsh` **once** after installing SkyPilot, and this will modify the `~/.zshrc` file accordingly. I've updated the docs to reflect this as well

Closes #887.

Tested:
- [x] bash
- [x] zsh
- [x] fish

### CODE REVIEW

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:363]: Can we make this idempotent? Otherwise, IIUC, if the user run the `sky --install-shell-completioin` will create multiple same lines in the `bashrc` file. Something similar to the trick in the following line can be applied.
https://github.com/skypilot-org/skypilot/blob/10967e6c17bee50827dfc6d1ea5cdd7fff2d6c8c/sky/templates/aws-ray.yml.j2#L95

Same for L366 and L370.

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:381]: UX: Can we add the actual commands for restarting the terminal as a hint for the user, e.g. `source ~/.bashrc` or `exec bash`?

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:978]: nit: It would be nice to have this `shell_complete` for sky launch as well, since sometimes the user want to rerun the `sky launch` for setup and file mounts.

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:361]: nit: can we place the file under `~/.sky/` instead of under `~` directly

File: docs/source/getting-started/installation.rst
Comment (Michaelvll) [inline docs/source/getting-started/installation.rst:165]: We need to have a way to uninstall it. [Reference for the terraform autocomplete](https://www.terraform.io/cli/commands#shell-tab-completion)

Also, can we automatically detect the shell the command is running in by default when the `zsh/bash/fish` is not provided?

**Review by Michaelvll — COMMENTED** [review]
Nice! Autocompletion is very useful for the users. I want this for a long time. Just tried it out and it seems the latency for completion is much shorter than last time I tried to implement it. Great work!

**Comment by concretevitamin** [comment]
This is awesome @iojw! Will let @Michaelvll review the details. High-level question: is it possible/easy to autocomplete the commands as well? E.g., `sky auto<tab>` would complete to `sky autostop`, or `sky st<tab>` would show either `status` or `stop`.

**Comment by iojw** [comment]
@concretevitamin Yup! It is already implemented in this PR. Completion for command names and option names will work out-of-the-box once shell completion is installed (works exactly as you mentioned).

File: sky/cli.py
Comment (iojw) [reply sky/cli.py:978]: To clarify, this is for the `--cluster` option for sky launch?

**Comment by romilbhardwaj** [comment]
This is awesome! Can we add this for storage names too? `sky storage delete <autocomplete>`

**Comment by iojw** [comment]
Thank you for the reviews and comments! 

I've made installation command idempotent, added an uninstallation command, and also added support for `sky storage`. I've also added auto shell detection if the user passes in `auto` to either option (e.g. `--install-shell-completion auto`). While ideally, we would be able to do this automatically if the user doesn't pass in any value at all (e.g. `--install-shell-completion`), it looks like `click` does not current support this flow as it requires every option to have an argument.

**Comment by romilbhardwaj** [comment]
Thanks for making the changes @iojw! I just tried out this PR and it's very nice to have autocomplete for cluster names and storage names. However, I faced some issues which prevent us from merging this PR in it's current state:

## 1. Autocompletion for command names is quite slow and ctrl+c produces a cryptic error
`sky sto<tab>` is very slow for me, taking up to 6-7 seconds to produce results. If I press `ctrl+c` after the command has executed, I get an error (likely because of an inflight autocomplete request that gets interrupted). See example:

[asciicast](https://asciinema.org/a/BUbsOSk4EkoOYXyN7TBffE24j)

Compared to other command line tools (e.g. kubectl) which are almost instantaneous:

[asciicast](https://asciinema.org/a/4LE0hDigj7nF6md4MC7fBTmkw)

I believe this is because our `cli.py` imports a ton of other libraries which slows down the autocomplete. We should either:
1. **Restructure our cli.py to be faster by making imports only when necessary**
OR 
2. **Use some other autocomplete mechanism?** E.g. instead of invoking click, can we have a fixed list of commands that the shell can quickly use?

I'm leaning towards option 2 here, but curious to hear what you and others have to say.

## 2. Adding this autocompletion breaks file name autocomplete
Say I have a file named `minimal.yaml` in my current dir. Typing `sky launch min<tab>` no longer autocompletes to `sky launch minimal.yaml`. Can we still support file name autocompletion while having command autocompletion?

**Comment by iojw** [comment]
@romilbhardwaj What machine are you running this on? I'm unable to reproduce the slow autocomplete - it always completes within 1-2 seconds for me:

[asciicast](https://asciinema.org/a/nu5JsWCfuEgYJZ6ic6wLwoUTX)

**Comment by romilbhardwaj** [comment]
Ah sorry I should have clarified - I'm on a 2018 Thinkpad X1 Carbon running SkyPilot on WSL. WSL slows down things a bit so I guess this might just be me.

On thinking more about it, maybe it's okay to merge this if the point 2 mentioned above about auto completing files is fixed. Since it's an optional feature, folks on slower setups like me can just keep it off. I'll let others chime in.

**Comment by iojw** [comment]
@romilbhardwaj I see! Yup, curious to see what others think 🙂. I've also updated the PR to support file name completion for `sky launch`. The issue is that we declare the `entrypoint` argument for `sky launch`  as a string, so `click` is unable to autocomplete for file names. To resolve this, I've added a custom shell completion function for file names.

**Comment by Michaelvll** [comment]
Hey @iojw, thank you for the fix! I just tried it out. The command and cluster name completion work quite well on my macbook, but the file name completion does not work. I typed the following lines in the root of the skypilot repo: `sky launch -c test-machine examp` and press tab. It took a long time and only add an additional `l` after the line. Did you observe that problem?

> On thinking more about it, maybe it's okay to merge this if the point 2 mentioned above about auto completing files is fixed. Since it's an optional feature, folks on slower setups like me can just keep it off. I'll let others chime in.

I agree with the point @romilbhardwaj  mentioned. By default, we should keep it off if that does not work well for all the users, but we also need to add the line in the doc saying something like "if you experience hanging after pressing tab, it is recommanded to turn it off by ...`.

**Comment by iojw** [comment]
@Michaelvll I'm unable to reproduce the issue - could you share more details about the directory structure and the setup you're using?

Here's a demo:
[asciicast](https://asciinema.org/a/8MY0MA7Kp1RDRTaqB0ODeDbWa)

**Comment by Michaelvll** [comment]
> @Michaelvll I'm unable to reproduce the issue - could you share more details about the directory structure and the setup you're using?

@iojw Thanks for trying it out! I tried it again and seems if the yaml file is in the current directory, the completion works well with 1-2 seconds latency, but if the yaml file is under in a directory, e.g. `examples/minimal.yaml` as I mentioned in my previous comment, the problem will happen, when you type `exa<tab>`.

**Comment by iojw** [comment]
@Michaelvll Got it - looks like the reason why this was happening is because I was using a custom completion that only looked at the current directory. I've figured out how to hook into the shell-provided completion system instead and updated the code to do this instead. It should work as expected now!

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:357]: nit: for readability, it would be better to rename this function to global_user_state.get_storage_names_start_with() so as to avoid ambiguity.

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:418]: nit: use red instead?

**Review by Michaelvll — APPROVED** [review]
Awesome! Thank you for the quick fix! Just tried out the completion and it became much faster and worked pretty well. I think it is good to go after fixing the following nits. :)

**Comment by iojw** [comment]
@Michaelvll Thanks! I've made the changes and will merge this in soon if there are no further comments. cc @romilbhardwaj @concretevitamin

**Review by romilbhardwaj — APPROVED** [review]
lgtm! This is a great improvement to our UX!

**Pull request merged**
